presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -226,4 +227,155 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
+            this.maximumValues = maximumValues;
+
+            // Create mutable version of encodingBits
+            List<Integer> bitLengths = new ArrayList<>(encodingBits);
+            this.dimensions = new int[maxBitLength];
+            for (int dimensionIndex = maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
+                for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
+                    if (bitLengths.get(bitLengthIndex) > 0) {
+                        dimensions[dimensionIndex]++;
+                        bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                    }
+                }
+            }
+        }
+
+        public List<Long> findAddresses()
+        {
+            addressRanges = new ArrayList<>();
+            findAddressesRecursively(maxBitLength - 1, 0, totalBitLength);
+            return addressRanges;
+        }
+
+        /**
+         * Recursive function to find all address ranges within given value ranges.
+         * <p/>
+         * Returns when the address range is either completely within or completely out of the value ranges.
+         * Recurses to a smaller Z-order curve if the address range only partially overlaps with the value ranges.
+         *
+         * @param level the level of recursion on the Z-curve TODO: elaborate
+         * @param startAddress the starting address of the current Z-curve level
+         * @param totalCurveBits the amount of bits in the current Z-curve level
+         */
+        private void findAddressesRecursively(int level, long startAddress, int totalCurveBits)
+        {
+            long endAddress = startAddress + ((long) 1 << totalCurveBits) - 1;
+
+            if (inRange(startAddress, endAddress)) {
+                addToAddressList(startAddress, endAddress);
+                return;
+            }
+
+            if (outOfRange(startAddress, endAddress)) {
+                return;
+            }
+
+            totalCurveBits -= dimensions[level];
+            int subtrees = 1 << dimensions[level];
+            for (int i = 0; i < subtrees; i++) {
+                findAddressesRecursively(level - 1, startAddress, totalCurveBits);
+                startAddress += ((long) 1 << totalCurveBits);
+            }
+        }
+
+        private void addToAddressList(long startAddress, long endAddress)
+        {
+            if ((addressRanges.size() == 0) || (startAddress != addressRanges.get(addressRanges.size() - 1) + 1)) {
+                // Add a new range
+                addressRanges.add(startAddress);
+                addressRanges.add(endAddress);
+            }
+            else {
+                // Combine ranges
+                addressRanges.set(addressRanges.size() - 1, endAddress);
+            }
+        }
+
+        private boolean inRange(long startAddress, long endAddress)
+        {
+            List<Integer> startValues = decode(startAddress);
```

Comment:
We probably want to extract this out so we don't need to call `decode` twice (once in `inRange` and once in `outOfRange`)?